cpp/include/raft/linalg/subtract.cuh

```diff
@@ -84,6 +88,137 @@ void subtractDevScalar(math_t* outDev,
   detail::subtractDevScalar(outDev, inDev, singleScalarDev, len, stream);
 }
 
+/**
+ * @defgroup sub Subtraction Arithmetic
+ * @{
+ */
+
+/**
+ * @brief Elementwise subtraction operation on the input buffers
+ * @tparam InType    Input Type raft::device_mdspan
+ * @tparam OutType   Output Type raft::device_mdspan
+ * @param handle raft::handle_t
+ * @param out    Output
+ * @param in1    First Input
+ * @param in2    Second Input
+ */
+template <typename InType,
+          typename OutType,
+          typename = raft::enable_if_device_mdspan<OutType, InType>>
+void subtract(const raft::handle_t& handle, InType in1, InType in2, OutType out)
```

Comment:
I notice the const missing here. Can we still declare the inputs `mdspan<const value_t>` somehow?

Reply:
@cjnolet Two issues here:

1. how to make sure that the inputs are `mdspan<const T>` for some `T`; and
2. the function should, but does not currently, constrain the mdspan all to have the same rank; and
3. the function should, but does not currently, constrain the mdspan to have all the same layouts.

Regarding (1), there are a few different ways to do this.  For example, we could constrain the parameters to be `mdspan` whose `element_type` is `const`.

```c++
template<class T>
constexpr bool is_const_mdspan_v = false;

template<class ElementType, class Extents, class Layout, class Accessor>
constexpr bool is_const_mdspan_v<mdspan<ElementType, Extents, Layout, Accessor>> =
  std::is_const_v<ElementType>;

template <typename InType,
  typename OutType,
  typename std::enable_if<
    is_device_mdspan<InType> &&
      is_device_mdspan<OutType> &&
      is_const_mdspan_v<InType>,
    bool>::type = true>
void subtract(const raft::handle_t& handle, InType in1, InType in2, OutType out);
```

This style is concise, and good choice of traits names can help document the interface.  However, use of SFINAE for overload resolution may also increase compilation time.  Another approach would be to overload with `mdspan` whose accessors are device accessible.  It's more verbose, but self-documenting, and it can avoid issues with traits classes needing to do `std::remove_const_t<std::remove_cv_t<T>>`.

```c++
template<class ElementType, class Extents, class Layout, class NestedAccessor, bool host_accessible>
using device_accessible_mdspan =
  mdspan<ElementType, Extents, Layout, host_device_accessor<NestedAccessor, host_accessible, true>>;

template<
  class InValueType1, class InExtents1, class InLayout1, class InNestedAccessor1, bool input1_host_accessible,
  class InValueType2, class InExtents2, class InLayout2, class InNestedAccessor2, bool input2_host_accessible,
  class OutValueType, class OutExtents, class OutLayout, class OutNestedAccessor, bool output_host_accessible>
void subtract(const raft::handle_t& handle,
  device_accessible_mdspan<const InValueType1, InExtents1, InLayout1, InNestedAccessor1, input1_host_accessible> in1,
  device_accessible_mdspan<const InValueType2, InExtents2, InLayout2, InNestedAccessor2, input2_host_accessible> in2,
  device_accessible_mdspan<OutValueType, OutExtents, OutLayout, OutNestedAccessor, output_host_accessible> out);
```

You might like to document that users should not provide template arguments explicitly.  That would give you implementation freedom to change the template parameters later, for example if you want to redefine `host_device_accessor` in the future.

Reply:
Thanks @mhoemmen! I agree that I don't see a reason these functions couldn't constrain the ins and outs to having the same layouts and indexing types, but your second example is exactly what we were trying to avoid with all of this (having to accept all the template parameters individually for each input).  I think your first example might not be a bad trade-off between forcing the inputs to be const qualified while still being agnostic of the layouts and indexing types and not bloating the number of template params. It also looks really clean, as you point out. 

Just out of curiosity, how much would you expect the SFINAE in your first example to slow down the compile time in comparison to the current approach (which just does the `enable_if_mdspan`?

Reply:
Why do we need to force the inputs to be const-qualified though? The interface they eventually call to translate to const pointers

Reply:
@mhoemmen @cjnolet instead of doing the device mdspan and const mdspan checks separately, I can just chain them together so we don't do extra work by having `enable_if_input_device_mdspan` and `enable_if_output_device_mdspan`. The first one enforces const-ness and the second one enforces the inverse